src/nfa/operations.cc

```diff
@@ -927,27 +927,52 @@ Nfa mata::nfa::algorithms::minimize_brzozowski(const Nfa& aut) {
     return determinize(revert(determinize(revert(aut))));
 }
 
-Nfa mata::nfa::minimize(
-                const Nfa& aut,
-                const ParameterMap& params)
-{
-    Nfa result;
-    // setting the default algorithm
-    decltype(algorithms::minimize_brzozowski)* algo = algorithms::minimize_brzozowski;
+Nfa mata::nfa::make_minimal_dfa(const Nfa& nfa, const ParameterMap& params) {
     if (!haskey(params, "algorithm")) {
         throw std::runtime_error(std::to_string(__func__) +
             " requires setting the \"algorithm\" key in the \"params\" argument; "
             "received: " + std::to_string(params));
     }
 
+    // Setting the algorithm. Default is Hopcroft.
     const std::string& str_algo = params.at("algorithm");
-    if ("brzozowski" == str_algo) {  /* default */ }
+    decltype(algorithms::minimize_brzozowski)* algo = algorithms::minimize_brzozowski;
+    if (str_algo == "brzozowski") { /* default */ }
     else {
+        throw std::runtime_error(std::string(__func__) +
+            " received an unknown value for the \"algorithm\" key: " + str_algo);
+    }
+
+    return algo(nfa);
+}
+
+Nfa mata::nfa::minimize(const Nfa &dfa, const ParameterMap& params)
+{
+    if (!haskey(params, "algorithm")) {
         throw std::runtime_error(std::to_string(__func__) +
-            " received an unknown value of the \"algorithm\" key: " + str_algo);
+            " requires setting the \"algorithm\" key in the \"params\" argument; "
+            "received: " + std::to_string(params));
+    }
+
+    // Setting the algorithm. Default is Hopcroft.
+    const std::string& str_algo = params.at("algorithm");
+    decltype(algorithms::minimize_hopcroft)* algo = algorithms::minimize_hopcroft;
+    if (str_algo == "hopcroft") {
+        /* default */;
+    } else if (str_algo == "brzozowski") {
+        algo = algorithms::minimize_brzozowski;
+    } else {
+        throw std::runtime_error(std::string(__func__) +
+            " received an unknown value for the \"algorithm\" key: " + str_algo);
     }
 
-    return algo(aut);
+    // Hopcroft algorithm does not work with non-trimmed automata.
+    const BoolVector is_used = dfa.get_useful_states();
+    bool is_trimmed = std::all_of(is_used.begin(), is_used.end(), [](bool b) { return b; });
```

Comment:
I would add this inside the if(hopcroft), so that it is not unnecessarily done for Brzozowski.